### ISSUE TITLE
Loading of files before training taking too much time.

### ISSUE THREAD

**Issue by alamnasim** [pyannote/pyannote-audio#247]
Hi @hbredin,
I am training pyannote on my own dataset and my train data contains total 634327 lines, 22143 npy files(60GB). I used use_**"memmap: False"** in config.yml. Now When I am training my model for Speaker_change detection it takes lots of time while loading npy files before training that i have shared below.

Please help me to rectify this error.

pyannote-change-detection train --gpu --to=500 --subset=train ${EXPERIMENT_DIR} odiosd.SpeakerDiarization.Hinglish
Training set: 0it [00:00, ?it/s]/home/ubuntu/anaconda3/envs/pyannote_work/lib/python3.6/site-packages/pyannote/database/util.py:391: UserWarning: "annotated" was approximated by "audio" duration.
  warnings.warn('"annotated" was approximated by "audio" duration.')
Training set: 16262it [55:37:23, 12.31s/it]
Training set: 498it [34:37:02, 339.75s/it]

Config.yml for speaker_change_detection:
task:
   name: SpeakerChangeDetection
   params:
      duration: 2.0      # sequences are 2s long
      collar: 0.100      # upsampling collar = 100ms
      non_speech: False  # do not try to detect non-speech/speaker changes
      batch_size: 128     # 64 sequences per batch
      per_epoch: 1       # one epoch = 1 day of audio

feature_extraction:
   name: Precomputed      # use MFCC from librosa
   params:
      root_dir: tutorials/feature_extraction
      use_memmap: False

Thanks.

**Comment by hbredin** [comment]
First, could you please provide some details on your dataset? 

```python
from pyannote.database import get_protocol
protocol = get_protocol('odiosd.SpeakerDiarization.Hinglish')
stats = protocol.stats('train')
print(f"Total number of speakers = {len(stats['labels'])}")
print(f"Total speech duration = {stats['annotation'] / 3600:.1f} hours")
print(f"Total annotated duration = {stats['annotated'] / 3600:.1f} hours")
```

**Comment by alamnasim** [comment]
Thanks for the quick reply.

My dataset details:
Total number of speakers = 32491
Total speech duration = 411.0 hours
Annotated uem part was not given in db.yml.

**Comment by alamnasim** [comment]
Hi @hbredin,
Could you please help me to resolve the above problem.
What should be the reason of this?

Thanks

**Comment by hbredin** [comment]
This is due to the combination of two things:
- the large size of your dataset (in terms of both number of speakers and speech duration)
- the fact that the code is not optimized for such large dataset

Short term solutions are:
- artificially reduce the number of speakers by labeling them A, B, C, D within each file (i.e. use the same label in each file even though speakers are different across files -- this is OK for training speaker change detection) 
- start with a smaller dataset

Long term solution is to rewrite SpeakerChangeDetectionGenerator to make it faster:
https://github.com/pyannote/pyannote-audio/blob/9ac612ee6b854a1a65c3d8992856550304969674/pyannote/audio/labeling/tasks/speaker_change_detection.py#L39

**Comment by hbredin** [comment]
I just pushed a major update to pyannote.audio in develop branch.

The batch generation process has been greatly optimized and this should solve your issue. Can you please try again by updating to the latest version available in develop branch and following the (updated) tutorials again?

In particular, there is now a --parallel=N option that makes the batch generator use (at most) that many background threads. You might not notice a big difference for the first few batches but you will eventually, once the batch generator reaches its cruising speed (it automagically adapts the number of threads after a little while). I'd be happy to hear feedback about it.

**Comment by alamnasim** [comment]
Thans a lot.

Yes the problem has solved, and files are loaded at quite fast rate like:
Training set: 16262it [5:32:38,  1.23s/it]

previously it was quite slow.
again thanks a lot.
